src/main/java/com/atlassian/bitbucket/jenkins/internal/trigger/register/BitbucketWebhookHandler.java

```diff
@@ -75,112 +79,118 @@ private void deleteWebhooks(List<BitbucketWebhook> webhooks) {
-            BitbucketWebhookRequest webhook = createRequest(request, event);
+        List<BitbucketWebhook> webhooks = webhookClient.getWebhooks().collect(Collectors.toList());
+        Set<BitbucketWebhook> serverSideWebhooks =
+                webhooks.stream()
+                        .filter(hook -> hook.getName().equals(request.getName()) && hook.getUrl().equals(callback))
+                        .collect(Collectors.toSet());
+
+        Set<BitbucketWebhookEvent> desiredEvents = new HashSet<>(events);
+        Set<BitbucketWebhookEvent> serverSideWebhookEvents = serverSideWebhooks.stream().flatMap(event ->
+                event.getEvents().stream().map(BitbucketWebhookEvent::findByEventId)).collect(Collectors.toSet());
+        desiredEvents.addAll(serverSideWebhookEvents);
+
+        if (serverSideWebhookEvents.containsAll(desiredEvents) &&
+                serverSideWebhookEvents.size() == desiredEvents.size()) {
+            //check that the webhooks are actually active and not just registered by disabled.
+            Optional<BitbucketWebhook> foundWebook = serverSideWebhooks.stream().filter(event -> event.isActive() &&
+                    callback.equalsIgnoreCase(event.getUrl()))
+                    .findFirst();
+            if (foundWebook.isPresent()) {
+                return foundWebook.get();
+            }
+        }
+
+        //Determining whether to create a new webhook
+        //We will handle mirror syncs separate from all other events.
+        // If no webhook with wanted events exist, we register a new one.
+        if (serverSideWebhooks.isEmpty()) {
+            BitbucketWebhookRequest webhook = createRequest(request, desiredEvents);
             BitbucketWebhook result = webhookClient.registerWebhook(webhook);
             LOGGER.info("New Webhook registered - " + result);
             return result;
         }
 
-        BitbucketWebhook mirrorSyncResult =
-                handleExistingWebhook(request, webhookWithMirrorSync, MIRROR_SYNCHRONIZED_EVENT);
-
-        BitbucketWebhook repoRefResult = handleExistingWebhook(request, webhookWithRepoRefChange, REPO_REF_CHANGE);
-
-        if (mirrorSyncResult != null && mirrorSyncResult.getEvents().contains(event.getEventId())) {
-            return mirrorSyncResult;
-        } else {
-            return repoRefResult;
-        }
-    }
-
-    @Nullable
-    private BitbucketWebhook handleExistingWebhook(WebhookRegisterRequest request,
-                                                   List<BitbucketWebhook> existingWebhooks,
-                                                   BitbucketWebhookEvent toSubscribe) {
-        BitbucketWebhook result = null;
-        if (existingWebhooks.size() > 0) {
-            result = update(existingWebhooks, request, toSubscribe);
-            existingWebhooks.remove(result);
-            deleteWebhooks(existingWebhooks);
-        }
-        return result;
+        return update(webhooks, request, desiredEvents);
```

Comment:
Should this be `update(serverSideWebhooks, request, desiredEvents)` instead? I have not tested this, but from reading the code, it seems that `webhooks` contains all the webhooks that Bitbucket Server has on this repository, and the `update` call will then delete the webhooks that were not set by Bitbucket Server Integration plugin.

Reply:
Hello @KalleOlaviNiemitalo

I will look at this to ensure we're not deleting any webhooks that are not "ours". Thank you for reporting this.

Reply:
It was a bug, I have put up a fix for it in https://github.com/jenkinsci/atlassian-bitbucket-server-integration-plugin/pull/273/files
thank you for reporting it!